### PR TITLE
ENH: add `FeatureMap[MAGtoContigs]` types and formats

### LINKED ISSUE #46: add `FeatureMap` and `FeatureMap[MAGtoContigs]` types
We will need the map to link contigs to unique MAGs right after binning.

```[tasklist]
### Tasks
- [ ] create `FeatureMap` and `FeatureMap[MAGtoContigs]` types
- [ ] create file format for `MAGtoContigs` type (this could be a json file mapping a unique MAG ID to a list of contig IDs)
```

### PR DESCRIPTION
**Pull request by misialq**
Closes #46.

### CODE REVIEW

File: q2_types_genomics/feature_map/tests/test_type.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class TestTypes(TestPluginBase):
+    package = "q2_types_genomics.feature_map.tests"
+
+    def test_feature_map_semantic_type_registration(self):
+        self.assertRegisteredSemanticType(FeatureMap)
+
+    def test_feature_map_mag_to_contigs_semantic_type_registration(self):
+        self.assertRegisteredSemanticType(MAGtoContigs)
+
+    def test_genome_data_genes_to_genes_dir_fmt_registration(self):
```
Comment (gregcaporaso) [inline q2_types_genomics/feature_map/tests/test_type.py:27]: Looks like a copy/paste error in the test function name.

**Review by gregcaporaso — CHANGES_REQUESTED** [review]
@misialq, this looks good to me. Just a few suggestions before merge.

File: q2_types_genomics/feature_map/tests/test_transformer.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+from q2_types_genomics.feature_map import MAGtoContigsFormat
+
+
+class TestTransformers(TestPluginBase):
+    package = "q2_types_genomics.feature_map.tests"
+
+    def setUp(self):
+        super().setUp()
+        self.valid_contig_map = {
```
[the remark this answers is not in the document]
Comment (misialq) [reply q2_types_genomics/feature_map/tests/test_transformer.py:21]: Hey @gregcaporaso, two thoughts here:
- I did not include a contig with a space in the name as it's impossible to have those in FASTA files; the only way for this to happen is when someone would edit the IDs by hand (which I find highly unlikely, considering the size of those datasets) - so my conclusion was not to include this check as it adds unnecessary processing of _a lot_ of IDs (for large datasets) - if you think, though, that it should be there  I'll add it in the validator
- similarly, I'm not testing whether there are duplicated IDs ad getting those is pretty much impossible either; the only scenario when it would happen is when user generated bins externaly using a bin refinement tool - the most common tools do make sure though that no contig ID is duplicated in each bin

This kind of comes down to the question, whether we want to safeguard against users importing data which was manually edited (and this, in the particular case of shotgun analysis is very unlikely, I guess)... What do you think? I'm happy to add the checks, but I don't want to add additional runtime whenever is not really necessary.

[the remark this answers is not in the document]
Comment (nbokulich) [reply q2_types_genomics/feature_map/tests/test_transformer.py:21]: yeah these sound like edge cases that are unlikely to ever occur.

[the remark this answers is not in the document]
Comment (gregcaporaso) [reply q2_types_genomics/feature_map/tests/test_transformer.py:21]: Sounds good - both of those suggestions were "if they were allowed", and it sounds like they are effectively not. I do still recommend varying the lengths of the lists of contigs.

**Review by gregcaporaso — APPROVED** [review]
